### PR TITLE
[WIP] Add BART for summarization training with CNN/DM using pytorch-lightning

### LINKED ISSUE #3004: BART : How can I train and evaluate BART on CNN/DM dataset
# ❓ Questions & Help

@sshleifer

I found examples for text summarization on CNN/DM dataset using BERT, but I couldn't find any example using BART.

**Are you going to add it later, or update the existing example to add BART, or it's not scheduled ?**

### PR DESCRIPTION
**Pull request by andr-ec**
This pull request adds to the example for BART for summarization. I used the [example for NER](https://github.com/huggingface/transformers/tree/master/examples/ner) using pytorch-lightning as guidance. This example will train on CNN/DM and evaluate, and get decent results, though I haven't trained it on the full dataset just yet. I'm sure there are better defaults for the hyperparams but these seem to work. 
I based this PR on the code I wrote in this [colab](https://colab.research.google.com/drive/1C4jEf0fnLiz6Xdx4TDz1OoO4BRCjCx1m).

This would hopefully close https://github.com/huggingface/transformers/issues/3004

## TODO
- [x] Be able to train the model on a GPU.
- [x] remove unused args
- [x] add test step and save results.  

Happy to hear any feedback!

### CODE REVIEW

**Comment by sshleifer** [comment]
Nice! @yjernite might be interested!

File: examples/summarization/bart/run_bart_sum.py
Comment (sshleifer) [inline examples/summarization/bart/run_bart_sum.py]: (naming) logits?

File: examples/summarization/bart/run_bart_sum.py
Comment (sshleifer) [inline examples/summarization/bart/run_bart_sum.py]: is this code identical to `training_step` can they both call a shared `_step` method?

File: examples/summarization/bart/run_bart_sum.py
Comment (sshleifer) [inline examples/summarization/bart/run_bart_sum.py]: (nit) commented code

File: examples/summarization/bart/run_bart_sum.py
Comment (sshleifer) [inline examples/summarization/bart/run_bart_sum.py]: (nit) why identical code twice? If not identical please comment :)

File: examples/summarization/bart/run_bart_sum.py
Comment (sshleifer) [inline examples/summarization/bart/run_bart_sum.py]: (nit) we want this or not?

File: examples/summarization/bart/run_bart_sum.py
Comment (sshleifer) [inline examples/summarization/bart/run_bart_sum.py]: 1024

File: examples/summarization/bart/run_bart_sum.py
Comment (sshleifer) [inline examples/summarization/bart/run_bart_sum.py]: already have batch_size (I know you didn't delete unused args yet)

File: examples/summarization/bart/run_bart_sum.py
Comment (sshleifer) [inline examples/summarization/bart/run_bart_sum.py]: lets delete this if it isn't actually supported.

File: examples/summarization/bart/utlis.py
Comment (sshleifer) [inline examples/summarization/bart/utlis.py]: (nit) typo in filename

File: examples/summarization/bart/utlis.py
Comment (sshleifer) [inline examples/summarization/bart/utlis.py]: why not block_size=1024?

File: examples/summarization/bart/utlis.py
Comment (sshleifer) [inline examples/summarization/bart/utlis.py]: why same thing twice?

**Review by sshleifer — DISMISSED** [review]
Thanks for starting this! I left a couple nitpicks, but looks reasonable to me. Were you planning on running finetuning for longer and posting results?

File: examples/summarization/bart/run_bart_sum.py
Comment (acarrera94) [reply examples/summarization/bart/run_bart_sum.py]: done

File: examples/summarization/bart/run_bart_sum.py
Comment (acarrera94) [reply examples/summarization/bart/run_bart_sum.py]: yes, done

File: examples/summarization/bart/run_bart_sum.py
Comment (acarrera94) [reply examples/summarization/bart/run_bart_sum.py]: removed

File: examples/summarization/bart/run_bart_sum.py
Comment (acarrera94) [reply examples/summarization/bart/run_bart_sum.py]: What identical code?

File: examples/summarization/bart/run_bart_sum.py
Comment (acarrera94) [reply examples/summarization/bart/run_bart_sum.py]: yes, updated

File: examples/summarization/bart/run_bart_sum.py
Comment (acarrera94) [reply examples/summarization/bart/run_bart_sum.py]: updated

File: examples/summarization/bart/run_bart_sum.py
Comment (acarrera94) [reply examples/summarization/bart/run_bart_sum.py]: correct, removed batch_size

File: examples/summarization/bart/run_bart_sum.py
Comment (acarrera94) [reply examples/summarization/bart/run_bart_sum.py]: done

File: examples/summarization/bart/utlis.py
Comment (acarrera94) [reply examples/summarization/bart/utlis.py]: done

File: examples/summarization/bart/utlis.py
Comment (acarrera94) [reply examples/summarization/bart/utlis.py]: done

File: examples/summarization/bart/utlis.py
Comment (acarrera94) [reply examples/summarization/bart/utlis.py]: removed

File: examples/summarization/bart/utlis.py
Comment (acarrera94) [reply examples/summarization/bart/utlis.py]: updated

**Comment by acarrera94** [comment]
I made those requested changes. And yes I'm planning to run finetuning this weekend and share results.  I only have access to a k80 so it'll take a while 🤷🏽‍♂️

**Comment by srush** [comment]
This looks awesome. Let's coordinate with https://github.com/huggingface/transformers/pull/3290 as well to share whatever code is possible.

**Comment by srush** [comment]
@nateraw can you do a review of this PR as well?

File: examples/summarization/bart/run_bart_sum.py
Comment (nateraw) [inline examples/summarization/bart/run_bart_sum.py]: This will eventually be handled by the transformer_base.py in #3290 , but for now can you add a config object in this `__init__`? Something like this:

```python
        config = BartConfig.from_pretrained(
            self.hparams.config_name if self.hparams.config_name else self.hparams.model_name_or_path,
            num_labels=num_labels,
            cache_dir=self.hparams.cache_dir if self.hparams.cache_dir else None,
        )
        tokenizer = BartTokenizer.from_pretrained(
            self.hparams.tokenizer_name if self.hparams.tokenizer_name else self.hparams.model_name_or_path,
            do_lower_case=self.hparams.do_lower_case,
            cache_dir=self.hparams.cache_dir if self.hparams.cache_dir else None,
        )
        model = BartForConditionalGeneration.from_pretrained(
            self.hparams.model_name_or_path,
            from_tf=bool(".ckpt" in self.hparams.model_name_or_path),
            config=config,
            cache_dir=self.hparams.cache_dir if self.hparams.cache_dir else None,
        )
        self.config, self.tokenizer, self.model = config, tokenizer, model
```

File: examples/summarization/bart/run_bart_sum.py
Comment (nateraw) [inline examples/summarization/bart/run_bart_sum.py]: Can we make this `self.model`?

File: examples/summarization/bart/run_bart_sum.py
Comment (nateraw) [inline examples/summarization/bart/run_bart_sum.py]: This pattern is being changed to `self(**)` from `self.forward(**)`

File: examples/summarization/bart/run_bart_sum.py
Comment (nateraw) [inline examples/summarization/bart/run_bart_sum.py]: Can we use the model's built in functionality for handling the loss (when you pass kwarg `lm_labels`), or no? I'm not familiar with the `BartForConditionalGeneration` class. 

From the [docstring](https://github.com/huggingface/transformers/blob/47591763137f17021928e686ef171f25c240f076/src/transformers/modeling_bart.py#L1053-L1059):
```
        tokenizer = BartTokenizer.from_pretrained('bart-large')
        model = BartForSequenceClassification.from_pretrained('bart-large')
        input_ids = torch.tensor(tokenizer.encode("Hello, my dog is cute",
        add_special_tokens=True)).unsqueeze(0)  # Batch size 1
        labels = torch.tensor([1]).unsqueeze(0)  # Batch size 1
        outputs = model(input_ids, labels=labels)
        loss, logits = outputs[:2]
```

File: examples/summarization/bart/run_bart_sum.py
Comment (nateraw) [inline examples/summarization/bart/run_bart_sum.py]: The transformer base class has a standardized setup for handling your dataloaders, and it might be a good idea to get yours lined up with that. The idea is:

1. You have a `prepare_data` function that will:
  a. process the data if a cache file is not found (no class attributes are set)
  b. do nothing if a cache file is found
2. A `load_dataset(self, mode: str, batch_size: int)` function that will take in a mode in `['train', 'dev', 'test']` and the corresponding batch size and returns a dataloader that has been loaded from the cached dataset. 

The base class will then have the functions `train_dataloader`, `val_dataloader`, and `test_dataloader` which look something like this:

```python
    def train_dataloader(self):
        train_batch_size = self.hparams.train_batch_size
        dataloader = self.load_dataset("train", train_batch_size)

        t_total = (
            (len(dataloader.dataset) // (train_batch_size * max(1, self.hparams.n_gpu)))
            // self.hparams.gradient_accumulation_steps
            * float(self.hparams.num_train_epochs)
        )
        scheduler = get_linear_schedule_with_warmup(
            self.opt, num_warmup_steps=self.hparams.warmup_steps, num_training_steps=t_total
        )
        self.lr_scheduler = scheduler
        return dataloader

    def val_dataloader(self):
        return self.load_dataset("dev", self.hparams.eval_batch_size)

    def test_dataloader(self):
        return self.load_dataset("test", self.hparams.eval_batch_size)
```

You could put the above lines in your code for now so when the base class is updated, you'll be able to update your example a bit faster by just removing them and letting the base class do the work.

**Review by nateraw — COMMENTED** [review]
Once #3290 gets merged, you'll have to update a few things, so I marked some here so you can get ahead of the curve on that. Once it's merged I'll be able to give a little more specific advice. Great work :+1:

**Comment by srush** [comment]
@acarrera94 I will try to get this working this week. If you are in the pytorch-lightning open slack we can also chat a bit more about the design.

File: examples/summarization/bart/run_bart_sum.py
Comment (acarrera94) [reply examples/summarization/bart/run_bart_sum.py]: I've made those changes, it looks like  #3290 has been merged now, how will this change?

File: examples/summarization/bart/run_bart_sum.py
Comment (acarrera94) [reply examples/summarization/bart/run_bart_sum.py]: What would saving the cached data look like? I can create instances of `CnnDailyMailDataset` in the `prepare_data` function as you described but how could I save the whole datasets themselves, and load them in the `load_dataset` function?

Comment (nateraw) [reply examples/summarization/bart/run_bart_sum.py]: This is probably already loading from cache in the backend, eh? Perhaps a non-issue then

**Comment by acarrera94** [comment]
@nateraw I've made all of those changes and it looks like #3290 has been merged, anything else that needs to change? Thanks!

**Comment by srush** [comment]
It's blocked on me, I should be able to get to it tonight.

File: examples/summarization/bart/run_bart_sum.py
Comment (yjernite) [inline examples/summarization/bart/run_bart_sum.py]: This asks the model to predict the input IDs, which leads to a loss of 0

I think you need:
```suggestion
        y_ids = y[:,:-1].contihuous()
        lm_labels = y[:,1:].clone()
        lm_labels[y[:,1:] == self.tokenizer.pad_token_id] = -100

        outputs = self(
            batch["source_ids"],
            attention_mask=batch["source_mask"],
            decoder_input_ids=y_ids,
            lm_labels=lm_labels,
        )
```

Comment (acarrera94) [reply examples/summarization/bart/run_bart_sum.py]: Good catch, thanks!

File: examples/summarization/bart/README.md
Comment (srush) [inline examples/summarization/bart/README.md]: Can you make this a bash script as well. Include the install command for pytorch-lightning as they need the latest version and it is not included in the `requirements.txt` file. Also you will need to add a line to put the base class in the path (following https://github.com/huggingface/transformers/pull/3290)

File: examples/summarization/bart/run_bart_sum.py
Comment (srush) [reply examples/summarization/bart/run_bart_sum.py]: So following https://github.com/huggingface/transformers/pull/3290 we would like you to use a shared base class for BartSystem. This should eliminate your `constructor`, `optimizer_step`, `configure_optimizers`, and several other functions.

File: examples/summarization/bart/run_bart_sum.py
Comment (srush) [inline examples/summarization/bart/run_bart_sum.py]: This part and below will all go away with the shared code.

File: examples/summarization/bart/run_bart_sum.py
Comment (srush) [inline examples/summarization/bart/run_bart_sum.py]: So this code actually won't work on distributed. Prepare data is only needed if you are downloading or pre-cacheing data. It only gets run by one of the workers. If this is loading from disk, then you should just do it in `train_dataloader` ...

**Review by srush — CHANGES_REQUESTED** [review]
Thanks! This code is nicely done. If you integrate with @nateraw 's work, I think it will eliminate about half the code. Also I would recommend just moving it to one file (utils will become very small.)

File: examples/summarization/bart/README.md
Comment (acarrera94) [reply examples/summarization/bart/README.md]: So you mean a bash script to preprocess the data? Since the user is downloading the dataset on their own, should we handle that by also downloading it for them as well in the bash script? And since the dataset is hosted on google drive, should we use `gdown` as a dependency? what are your thoughts?

Comment (srush) [reply examples/summarization/bart/README.md]: It's okay if you first tell them to download and preprocess the data separately. Just want to make sure the dependencies install and model run are in a script.

File: examples/summarization/bart/run_bart_sum.py
Comment (srush) [inline examples/summarization/bart/run_bart_sum.py]: Don't think you need these guys anymore

**Review by srush — APPROVED** [review]

**Comment by srush** [comment]
New code looks great. Excited to try it out!

File: examples/summarization/bart/run_bart_sum.py
Comment (acarrera94) [reply examples/summarization/bart/run_bart_sum.py]: We do otherwise we get `AttributeError: 'Namespace' object has no attribute 'learning_rate'` since these are model specific args and aren't part of `add_generic_args`

Comment (srush) [reply examples/summarization/bart/run_bart_sum.py]: Oh, you should just call the parent `add_model_specific_args` function before adding the BART args
https://github.com/huggingface/transformers/blob/master/examples/transformer_base.py#L148

They are mostly all here:
https://github.com/huggingface/transformers/blob/master/examples/transformer_base.py#L181

File: examples/summarization/bart/utils.py
Comment (srush) [inline examples/summarization/bart/utils.py]: Minor: What if we just call this: "SummarizationDataset" , and remove mention of CNN/DM, which is just one specific dataset

Comment (acarrera94) [reply examples/summarization/bart/utils.py]: Done!

**Comment by srush** [comment]
Thanks for sticking with it @acarrera I'm really impressed how concise this became. Next we can get some numbers.

**Comment by sshleifer** [comment]
@acarrera94 `run_train.sh` is using 19GB on my system.
Does your system use less?
I am also seeing no memory savings from adding `--fp16`. 
Thanks!

**Comment by acarrera94** [comment]
@sshleifer I usually ran it using --max_seq_lengt=756. And that used less than 16gb of ram with a batch size of 4, so we might want to change that default.  And I haven’t tried it using --fp16. That comes from BaseTransformer right?

**Pull request merged**
